src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -930,7 +932,21 @@ public static IGraphQLBuilder ConfigureExecutionOptions(this IGraphQLBuilder bui
         /// </remarks>
         public static IGraphQLBuilder ConfigureExecutionOptions(this IGraphQLBuilder builder, Func<ExecutionOptions, Task> action)
         {
+#pragma warning disable CS0618 // Type or member is obsolete
             builder.Services.Register<IConfigureExecutionOptions>(new ConfigureExecutionOptions(action ?? throw new ArgumentNullException(nameof(action))));
+#pragma warning restore CS0618 // Type or member is obsolete
+            return builder;
+        }
+
+        /// <summary>
+        /// Configures an action that can modify or replace document execution behavior.
+        /// </summary>
+        /// <remarks>
+        /// <see cref="ExecutionOptions.RequestServices"/> can be used within the delegate to access the service provider for this execution.
+        /// </remarks>
+        public static IGraphQLBuilder ConfigureExecution(this IGraphQLBuilder builder, Func<Func<ExecutionOptions, Task<ExecutionResult>>, ExecutionOptions, Task<ExecutionResult>> action)
```

Comment:
```suggestion
        public static IGraphQLBuilder ConfigureExecution(this IGraphQLBuilder builder, Func<ExecutionOptions, ExecutionDelegate, Task<ExecutionResult>> action)
```

Reply:
@Shane32 Compare this with
```cs
 /// <summary>
        /// Adds a middleware delegate defined in-line to the application's request pipeline.
        /// If you aren't calling the next function, use <see cref="RunExtensions.Run(IApplicationBuilder, RequestDelegate)"/> instead.
        /// </summary>
        /// <param name="app">The <see cref="IApplicationBuilder"/> instance.</param>
        /// <param name="middleware">A function that handles the request and calls the given next function.</param>
        /// <returns>The <see cref="IApplicationBuilder"/> instance.</returns>
        public static IApplicationBuilder Use(this IApplicationBuilder app, Func<HttpContext, RequestDelegate, Task> middleware)
        {
            return app.Use(next => context => middleware(context, next));
        }
```

`ConfigureExecution` <-> `Use`
`IGraphQLBuilder` <-> `IApplicationBuilder`
`action` <-> `middleware`
`ExecutionOptions` <-> `HttpContext`
`ExecutionDelegate` <-> `RequestDelegate`
`Task<ExecutionResult>` <-> `Task`

Suggestions from me:

1. Maybe rename method name and argument name to align with ASP.NET stuff? I'm 50/50.
2. Add other overloads from ASP.NET: 
` public static IApplicationBuilder Use(this IApplicationBuilder app, Func<HttpContext, Func<Task>, Task> middleware)`
and
`  IApplicationBuilder Use(Func<RequestDelegate, RequestDelegate> middleware);`
3. Add docs mentioning that GraphQL.NET document execution pipeline follows the very same design as ASP.NET http request pipeline.

Reply:
ASP.NET Core docs discourage the use of the first overload you suggested, so I would suggest we simply skip that one.  The second overload requires a change in design to implement: it's an application builder, so instead of `IConfigureExecution` we would have `IExecutionBuilder` (or something named like that), where the interface method is `ExecutionDelegate BuildExecution(ExecutionDelegate next)`.

Do you think it's worth it?  Ideally, the start of the execution would initialize an `ExecutionContext` (similar to `HttpContext`), and everything - document parsing, document caching, document listeners, document validation, and even the execution strategy selection and execution - would be based on installable middleware, similar to ASP.NET Core.  However, we would need a place to configure the builder pipeline in case of branches, as pulling a list from DI would be insufficient.  I feel like that's too big of a change to tackle for v5, although it is worth considering for v6.

I would suggest we leave this a bit simpler for v5, tackling the ability to add middleware around the existing document executer, and if we wish to rewrite the execution pipeline from scratch, we do so for v6.

What do you think?  If you still think we should make the changes you proposed, fine, but can you be a little more specific as to the interface definition that should be added, and builder methods to be added?

Example interface:

```cs
// now:
public interface IConfigureExecution
{
    Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next);
}
public static IGraphQLBuilder ConfigureExecution(this IGraphQLBuilder builder, Func<ExecutionOptions, ExecutionDelegate, Task<ExecutionResult>> action);

// suggestion???
public interface IExecutionBuilder
{
    ExecutionDelegate Build(ExecutionDelegate next);
}
public static IGraphQLBuilder ConfigureExecution(this IGraphQLBuilder builder, Func<ExecutionOptions, ExecutionDelegate, Task<ExecutionResult>> middleware);
public static IGraphQLBuilder ConfigureExecution(this IGraphQLBuilder builder, Func<ExecutionDelegate, ExecutionDelegate> middleware);
```

Again, personally I think I'd just leave it as shown in this PR for v5.

One other option: we could leave the interface as-is and add the extra configuration method like shown below, but I don't like it - it seems like a hack.

```cs
public static IGraphQLBuilder ConfigureExecution(this IGraphQLBuilder builder, Func<ExecutionDelegate, ExecutionDelegate> middleware)
{
    builder.Services.Register<IConfigureExecution>(new ConfigureExecution((options, next) => middleware(next)(options)));
    return builder;
}
```

Reply:
@sungam3r bump - if we answer this one question, we may be able to merge this now or shortly.

Reply:
I agree to postpone redesign to v6.

Reply:
#3131